pkg/clustermesh/common/namespacewatcher/watcher.go

```diff
@@ -0,0 +1,480 @@
+func NewAlwaysGlobalTracker() GlobalNamespaceTracker {
+	return &alwaysGlobalTracker{}
+}
+
+// alwaysGlobalTracker is used for backward compatibility when namespace watching is not enabled
+type alwaysGlobalTracker struct{}
+
+func (d *alwaysGlobalTracker) IsGlobalNamespace(namespace string) bool        { return true }
+func (d *alwaysGlobalTracker) GetGlobalNamespaces() sets.Set[string]          { return sets.New[string]() }
+func (d *alwaysGlobalTracker) RegisterProcessor(processor NamespaceProcessor) {}
+func (d *alwaysGlobalTracker) IsFilteringActive() bool                        { return false }
+func (d *alwaysGlobalTracker) IsGlobalService(obj interface {
+	GetAnnotations() map[string]string
+	GetNamespace() string
+}) bool {
+	return annotation.GetAnnotationIncludeExternal(obj)
+}
+func (d *alwaysGlobalTracker) IsSharedGlobalService(obj interface {
+	GetAnnotations() map[string]string
+	GetNamespace() string
+}) bool {
+	return d.IsGlobalService(obj) && annotation.GetAnnotationShared(obj)
+}
+
+// GlobalNamespaceTracker tracks which namespaces are marked as global for ClusterMesh export.
+type GlobalNamespaceTracker interface {
+	// IsGlobalNamespace returns true if the given namespace should be exported globally.
+	// Returns true by default for backwards compatibility if no specific configuration exists.
+	IsGlobalNamespace(namespace string) bool
+
+	// GetGlobalNamespaces returns the set of namespaces that are currently marked as global.
+	GetGlobalNamespaces() sets.Set[string]
+
+	// RegisterProcessor registers a processor that will be notified when namespace status changes
+	RegisterProcessor(processor NamespaceProcessor)
+
+	// IsFilteringActive returns true if namespace-based filtering is currently active
+	IsFilteringActive() bool
+
+	// IsGlobalService checks if a service is global considering both
+	// the service annotation and namespace filtering when active.
+	IsGlobalService(obj interface {
+		GetAnnotations() map[string]string
+		GetNamespace() string
+	}) bool
+
+	// IsSharedGlobalService checks if a service is both global and shared
+	IsSharedGlobalService(obj interface {
+		GetAnnotations() map[string]string
+		GetNamespace() string
+	}) bool
+}
+
+// NamespaceProcessor handles processing when namespace global status changes
+type NamespaceProcessor interface {
+	ProcessNamespaceChange(namespace string, isGlobal bool)
+	// GetAllNamespaces returns all namespaces that have resources in this processor's resource index
+	GetAllNamespaces() []string
+}
+
+const (
+	// OptClusterMeshDefaultGlobalNamespace is the name of the clustermesh-default-global-namespace option
+	OptClusterMeshDefaultGlobalNamespace = "clustermesh-default-global-namespace"
+)
+
+// Config configures the namespace watcher behavior.
+type Config struct {
+	// DefaultGlobalNamespace determines the default behavior for namespaces when
+	// namespace-based filtering is active (i.e., when at least one namespace is annotated).
```

Comment:
In the CFP we agreed that the behavior is controlled by the option only, and not by the namespace annotations. Checking the annotations adds unnecessary complexity, and can lead to surprises due to a single change completely changing the behavior globally.

Reply:
> In the CFP we agreed that the behavior is controlled by the option only, and not by the namespace annotations.

Agree. Removed the annotation based logic.